pygfx/resources/_base.py

```diff
@@ -0,0 +1,77 @@
+import weakref
+
+from ..utils.trackable import Trackable
+
+
+STRUCT_FORMAT_ALIASES = {"c": "B", "l": "i", "L": "I"}
+
+FORMAT_MAP = {
+    "b": "i1",
+    "B": "u1",
+    "h": "i2",
+    "H": "u2",
+    "i": "i4",
+    "I": "u4",
+    "e": "f2",
+    "f": "f4",
+}
+
+
+class Resource(Trackable):
+    """Resource base class."""
+
+    def __init__(self):
+        super().__init__()
+        registry._gfx_register(self)
+
+    def _gfx_mark_for_sync(self):
+        registry._gfx_mark_for_sync(self)
+
+
+class ResourceRegistry:
+    """Singleton registry for resources."""
+
+    def __init__(self):
+        self._all = weakref.WeakSet()
+        self._syncable = weakref.WeakSet()
+
+    def _gfx_register(self, resource):
+        """Add a resource to the registry."""
+        if not isinstance(resource, Resource):
+            raise TypeError("Given object is not a Resource")
+        self._all.add(resource)
+
+    def _gfx_mark_for_sync(self, resource):
```

Comment:
I don't really see why these methods on registry need to be private?

Reply:
It's more that I don't see a reason to make them public, so a private method seems safer (can make it public later if needed, but the reverse is less true).